R/decks.R

```diff
@@ -0,0 +1,517 @@
+    }
+
+    if (missing(file)) {
+        file <-  paste0(name(deck), ext)
+    }
+    crDownload(dl_link, file)
+}
+
+setMethod("length", "CrunchDeck", function(x){
+    return(length(slides(x)))
+})
+
+setMethod("cubes", "CrunchDeck", function(x){
+    out <- lapply(seq_len(length(x)), function(i){
+        cubes <- cubes(x[[i]])
+        if (length(cubes) == 1) {
+            cubes <- cubes[[1]]
+        }
+        return(cubes)
+    })
+    names(out) <- titles(x)
+    return(out)
+})
+
+setMethod("show", "CrunchDeck", function(object){
+    print(cubes(object))
+})
+
+# Slide Catalog -----------------------------------------------------------
+
+setMethod("initialize", "SlideCatalog", function(.Object, ...){
+    .Object <- callNextMethod()
+    order <- crGET(.Object@orders$flat)
+    order <- unlist(order$graph)
+    .Object@index <- .Object@index[match(order, names(.Object@index))]
+    return(.Object)
+})
+
+setMethod("names", "SlideCatalog", function(x) titles(x))
+setMethod("names<-", "SlideCatalog", function(x, value) titles(x) <- value)
+
+setMethod("titles", "SlideCatalog", function (x){
+    as.character(lapply(index(x), function(a) a$title))
+})
+
+setMethod("titles<-", "SlideCatalog", function (x, value){
+    updateSlideCatTitles(x, value, "title")
+})
+
+setMethod("subtitles", "SlideCatalog", function (x){
+    as.character(lapply(index(x), function(a) a$subtitle))
+})
+
+setMethod("subtitles<-", "SlideCatalog", function (x, value){
+    updateSlideCatTitles(x, value, "subtitle")
+})
+
+updateSlideCatTitles <- function(x, value, type) {
+    # TODO update a single title.
+    stopifnot(is.character(value))
+    payload <- mapply(function (item, new_title){
+        out <- item[type]
+        out[[type]] <- new_title
+        return(out)
+    }, x@index, value, SIMPLIFY = FALSE)
+    crPATCH(self(x), body = toJSON(payload))
+    invisible(refresh(x))
+}
+
+setMethod("[[", "SlideCatalog", function (x, i, ...) {
+  getEntity(x, i, CrunchSlide)
+})
+
+setMethod("[[<-", c("SlideCatalog", "numeric", "missing", "CrunchSlide"),
+          function(x, i, j, value) {
+              if (length(i) > 1) {
+                  #TODO, allow assignment of more than one slide
+              }
+              if (i > length(x) + 1) {
+                  #TODO what to do with missing slide entries
+                  i <- length(x) + 1
+              }
+
+              n_slides <- length(x)
+
+              payload <- value@body[c("title", "subtitle")]
+              payload$analyses <- analysesToQueryList(analyses(value))
+              payload <- wrapEntity(body = payload)
+              crPOST(self(x), body = toJSON(payload))
+
+              if (i < n_slides) {
+                  # You can't modify the contents of a slide by patching it
+                  # so we need to add the new slide, delete the original slide,
+                  # and reorder the slideCatalog.
+                  new_order <- moveLastElement(seq_len(n_slides + 1), i)
```

Comment:
Is there a better way to accomplish this? The only other method I could think of was updating the title and subtitle of the slide, then pulling the analysis catalog and updating that, but that seems even clunkier.